core/src/main/java/org/infinispan/marshall/core/GlobalMarshaller.java

```diff
@@ -370,8 +414,189 @@ private void writeNonNullableObject(Object obj, BytesObjectOutput out) throws IO
       }
    }
 
+   private void writeArray(Class<?> clazz, Object array, BytesObjectOutput out) throws IOException {
+      out.writeByte(ID_ARRAY);
+      Class<?> componentType = clazz.getComponentType();
+
+      int length = Array.getLength(array);
+      boolean singleType = true;
+      Class<?> elementType = null;
+      int flags;
+      if (length == 0) {
+         flags = FLAG_ARRAY_EMPTY;
+      } else {
+         if (length <= Primitives.SMALL_ARRAY_MAX) {
+            flags = FLAG_ARRAY_SMALL;
+         } else if (length <= Primitives.MEDIUM_ARRAY_MAX) {
+            flags = FLAG_ARRAY_MEDIUM;
+         } else {
+            flags = FLAG_ARRAY_LARGE;
+         }
+         Object firstElement = Array.get(array, 0);
+         if (firstElement != null) {
+            elementType = firstElement.getClass();
+         }
+         for (int i = 1; i < length; ++i) {
```

Comment:
I'm not sure compaction should really be our goal. Compaction helps reduce payload sizes but if the code becomes to complex for the CPUs to be able to predict what will happen and get ahead of it, or for us to understand the code, I'm not sure about its merits, particularly if there's no hard data to back up one strategy versus the other.